pkg/env/env.go

```diff
@@ -387,38 +433,116 @@ func (e *testEnv) getFinishActions() []action {
 	return e.getActionsByRole(roleFinish)
 }
 
-func (e *testEnv) execFeature(ctx context.Context, t *testing.T, featName string, f types.Feature) context.Context {
-	// feature-level subtest
-	t.Run(featName, func(t *testing.T) {
-		// skip feature which matches with --skip-feature
-		if e.cfg.SkipFeatureRegex() != nil && e.cfg.SkipFeatureRegex().MatchString(featName) {
-			t.Skipf(`Skipping feature "%s": name matched`, featName)
-		}
-
-		// skip feature which does not match with --feature
-		if e.cfg.FeatureRegex() != nil && !e.cfg.FeatureRegex().MatchString(featName) {
-			t.Skipf(`Skipping feature "%s": name not matched`, featName)
+func (e *testEnv) runTests() bool {
+	return !e.cfg.DryRunMode() && !e.gatherMode
+}
+
+func (e *testEnv) shoulFeatureOrAssessmentBeRun(f types.Feature, a types.Step, assessmentIndex int) (skipped bool, message string) {
+	var requiredRegexp *regexp.Regexp
+	var skipRegexp *regexp.Regexp
+	var kind string
+	var name string
+	if f != nil {
+		kind = "feature"
+		requiredRegexp = e.cfg.FeatureRegex()
+		skipRegexp = e.cfg.SkipFeatureRegex()
+		name = f.Name()
+	} else {
+		kind = "assessment"
+		requiredRegexp = e.cfg.AssessmentRegex()
+		skipRegexp = e.cfg.SkipAssessmentRegex()
+		name = a.Name()
+		if name == "" {
+			name = fmt.Sprintf("Assessment-%d", assessmentIndex)
 		}
+	}
 
+	if requiredRegexp != nil && !requiredRegexp.MatchString(name) {
+		skipped = true
+		message = fmt.Sprintf(`Skipping %s "%s": name not matched`, kind, name)
+		return
+	}
+	if skipRegexp != nil && skipRegexp.MatchString(name) {
+		skipped = true
+		message = fmt.Sprintf(`Skipping %s: "%s": name matched`, kind, name)
+		return
+	}
+	if f != nil {
 		// skip if labels does not match
 		// run tests if --labels values matches the feature labels
 		for k, v := range e.cfg.Labels() {
 			if f.Labels()[k] != v {
-				t.Skipf(`Skipping feature "%s": unmatched label "%s=%s"`, featName, k, f.Labels()[k])
+				skipped = true
+				message = fmt.Sprintf(`Skipping feature "%s": unmatched label "%s=%s"`, name, k, f.Labels()[k])
+				return
 			}
 		}
 
 		// skip running a feature if labels matches with --skip-labels
 		for k, v := range e.cfg.SkipLabels() {
 			if f.Labels()[k] == v {
-				t.Skipf(`Skipping feature "%s": matched label provided in --skip-lables "%s=%s"`, featName, k, f.Labels()[k])
+				skipped = true
+				message = fmt.Sprintf(`Skipping feature "%s": matched label provided in --skip-lables "%s=%s"`, name, k, f.Labels()[k])
+				return
 			}
 		}
+	}
+	return
+}
+
+func (e *testEnv) updateTestStats(m *testing.M, kind string, mode string, count int) {
```

Comment:
Yes. I am fine with splitting this for sure. I was taking care of the stats with the same PR since @johnSchnake  has suggested we should include the stats as well when discussing the orginal thread that introduced the ask for a `dry-run` flag. But you are right. It gets complicated to review/merge as well if we keep these two things together. +1 to moving this to a different PR/